### ISSUE TITLE
Add support for composite templates.

### ISSUE THREAD

**Issue by gm666q** [gtk-rs/gtk#938]
Add support for using composite templates from Builder/Glade.

**Comment by cyclopsian** [comment]
I was able to get this to work with some hacking and unsafe blocks. It requires a patch to the glib bindings to allow running code in the instance initializer after the data from `new()` gets moved into the private struct: [instance_init_0.9.3 patch](https://github.com/gtk-rs/glib/compare/0.9.3...cyclopsian:instance_init_0.9.3). Incomplete sample code is below. Requesting design advice from someone who knows a bit more about wrapping unsafe APIs.

mywindow.rs:
```rust
#[macro_use]
extern crate memoffset;

#[derive(Debug)]
pub struct MyWindowPrivate {
    button: *mut gtk_sys::GtkButton
}

impl ObjectSubclass for MyWindowPrivate {
    const NAME: &'static str = "MyWindow";
    type ParentType = gtk::ApplicationWindow;
    type Instance = subclass::simple::InstanceStruct<Self>;
    type Class = subclass::simple::ClassStruct<Self>;

    glib_object_subclass!();

    fn class_init(klass: &mut subclass::simple::ClassStruct<Self>) {
        let template = glib::Bytes::from_static(include_bytes!("mywindow.glade"));
        let object_class = klass as *mut _ as *mut gobject_sys::GObjectClass;
        let type_class = object_class as *mut gobject_sys::GTypeClass;

        unsafe {
            let data = Self::type_data();
            let offset = data.as_ref().private_offset;
            let widget_class =
                gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
                as *mut gtk_sys::GtkWidgetClass;

            gtk_sys::gtk_widget_class_set_template(widget_class, template.to_glib_none().0);
            gtk_sys::gtk_widget_class_bind_template_child_full(
                widget_class,
                "mybutton".to_glib_none().0,
                false as glib_sys::gboolean,
                offset + (offset_of!(MyWindowPrivate, button) as isize));
        }
    }

    fn new() -> Self {
        Self { button: std::ptr::null_mut() }
    }

    fn instance_init(&mut self, _: &subclass::simple::ClassStruct<Self>) {
        self.get_instance().init_template();
    }
}

impl ObjectImpl for MyWindowPrivate {
    glib_object_impl!();
    
    fn constructed(&self, obj: &glib::Object) {
        self.parent_constructed(obj);

        let button = unsafe {
            Option::<gtk::Button>::from_glib_none(self.button).unwrap()
        };
        button.set_label("hello");
    }
}
```

mywindow.ui:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<interface>
  <requires lib="gtk+" version="3.20"/>
  <template class="MyWindow" parent="GtkApplicationWindow">
    <child>
      <object class="GtkButton" id="mybutton">
        <property name="label" translatable="yes">button</property>
      </object>
    </child>
  </template>
</interface>
```

**Comment by sdroege** [comment]
Thanks, that's very interesting :) I'll try to find some time to go through this in more detail some time this week.

**Comment by cyclopsian** [comment]
Ideally it would have some magic so that `gtk::Button` could be used in the struct instead of `gtk_sys::GtkButton`. The pointer would get need to get turned in-place into an `ObjectRef` after the `init_template` call. There probably needs to be a custom wrapper type for this, i.e. not `Option` or `OnceCell`.

**Comment by cyclopsian** [comment]
This seems to work OK for me, for now:

```rust
#[derive(Debug)]
pub struct TemplateWidget<T>
    where T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>
{
    ptr: *mut <T as ObjectType>::GlibType
}

impl<T> TemplateWidget<T>
    where T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>
{
    fn new() -> Self {
        Self { ptr: std::ptr::null_mut() }
    }
    fn unwrap(&self) -> T {
        unsafe {
            Option::<T>::from_glib_none(self.ptr).unwrap()
        }
    }
}

#[derive(Debug)]
pub struct MyWindow {
    button: TemplateWidget<gtk::Button>
}
```

**Comment by sdroege** [comment]
> after the data from `new()` gets moved into the private struct: [instance_init_0.9.3 patch](https://github.com/gtk-rs/glib/compare/0.9.3...cyclopsian:instance_init_0.9.3)

Why is this needed? I assume for `gtk_widget_init_template()`? We might want to add something generic for this so that in `src/subclass/widget.rs` (similar to the `override_vfuncs`) this is called automatically if the class has a template. Maybe a second optional (`unsafe`!) function in `IsSubclassable` that gets the instance as a raw pointer. What do you think?

This also will have to be an `unsafe` function as at this point the object is not "complete" yet and random crashes might happen because subclasses were not initialized yet.

Similarly I think the `bind_template_child()` and `bind_template_callback()` bindings will have to be `unsafe`. We could abstract the whole thing into some macro for making its usage safe probably, but I don't see how exposing these functions themselves can be done in a safe way.

As a safe solution for now you could use `gtk_widget_get_template_child()`, which will give you a properly wrapped Rust object that you can then downcast to the given target type, safely.

`gtk_widget_class_set_template()` seems like something that can be easily and safely be added to an `WidgetClassSubclassExt` extension trait. This is very similar to what is in the GStreamer bindings [here](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/962da8a63d1b61bbdf44f4486f55fb6c13864c36/gstreamer/src/subclass/element.rs#L281).

So far my thoughts about this, let me know what you think :)

**Comment by ebassi** [comment]
> As a safe solution for now you could use `gtk_widget_get_template_child()`, which will give you a properly wrapped Rust object that you can then downcast to the given target type, safely.

Bindings for dynamic languages do this, as of course they can't access the instance structure, both public or private. The main thing you want to avoid is having to call `gtk_widget_get_template_child()` more than once, as that's a hash table lookup and a bunch of type checks.

**Comment by sdroege** [comment]
> The main thing you want to avoid is having to call `gtk_widget_get_template_child()` more than once

Yeah, I think you'd call that in `GObject::constructed` and fill in your private struct with the widgets as needed then (using `OnceCell`s, for example).

**Comment by cyclopsian** [comment]
>Why is this needed? I assume for gtk_widget_init_template()?

Yes, that is correct. It probably should be called automatically if `gtk_widget_class_set_template()` is called at all.

I tried with `gtk_widget_get_template_child()`, the problem with that is the class still needs to call `gtk_widget_class_bind_template_child_full()` first in order to place the widget in the hash table, which means there still needs to be extra space allocated for it in the private struct. Ultimately I don't know how much more safe that would be.

Current attempt at writing an incredibly basic widget subclass is here: [widget_subclass.rs](https://gist.github.com/cyclopsian/2c3b5cc03f5214e2bddc527be11a4816). The types of bound widgets will only be verified when `TemplateWidget::unwrap()` is called.

I didn't make a wrapper for `gtk_widget_class_bind_template_callback_full()` because there doesn't seem to be any way there to verify the type signature of any given `GCallback` before `init_template()` is called. My experience with that is that I had to wait until the signal is emitted and then the application will just crash if the function has the wrong arguments.

An example implementor now looks like this:

```rust
#[derive(Debug, Default)]
pub struct MyWindowPrivate {
    button: TemplateWidget<gtk::Button>
}

impl WidgetSubclass for MyWindowPrivate {}

impl ObjectSubclass for MyWindowPrivate {
    const NAME: &'static str = "MyWindow";
    type ParentType = gtk::ApplicationWindow;
    type Instance = subclass::simple::InstanceStruct<Self>;
    type Class = subclass::simple::ClassStruct<Self>;

    glib_object_subclass!();

    fn class_init(klass: &mut subclass::simple::ClassStruct<Self>) {
        Self::set_template_static(klass, include_bytes!("mywindow.glade"));
        Self::bind_template_child_private(klass, "mybutton",
                                          offset_of!(Self => button));
    }

    fn new() -> Self {
        Self { ..Default::default() }
    }

    fn instance_init(&mut self, _: &subclass::simple::ClassStruct<Self>) {
        self.get_instance().init_template();
    }
}
```

**Comment by sdroege** [comment]
> I tried with `gtk_widget_get_template_child()`, the problem with that is the class still needs to call `gtk_widget_class_bind_template_child_full()` first in order to place the widget in the hash table, which means there still needs to be extra space allocated for it in the private struct. Ultimately I don't know how much more safe that would be.

That sounds unexpected. @ebassi can you confirm that `get_template_child()` requires things to be bound first somehow?

> Yes, that is correct. It probably should be called automatically if `gtk_widget_class_set_template()` is called at all.

Great, I'd suggest as a first step to add something along the lines of what I wrote above to the bindings. Something to hook into `instance_init()` so that this can be automatically called, and safe bindings for `gtk_widget_class_set_template()` (and the resource variant of it).

Those are the easy parts that we should get out of the way and then we can think about the other pieces :)

> I didn't make a wrapper for `gtk_widget_class_bind_template_callback_full()` because there doesn't seem to be any way there to verify the type signature of any given `GCallback` before `init_template()` is called. My experience with that is that I had to wait until the signal is emitted and then the application will just crash if the function has the wrong arguments.

That's correct, and unfortunately it also does not work with a `GClosure` but directly with a function pointer so we can't do any runtime checks like with the `GtkBuilder` API either. In GTK4 this works completely different though, so hopefully less of a problem there.

Just means that for GTK3 we'd have to connect signals manually in e.g. `constructed()`. Could be worse!

> An example implementor now looks like this:

That actually looks quite nice and like something we could get in as a second step after the above. I'm not 100% sure about the safety of the `bind_template_child_private()` part, I'll review that in more detail once there's a MR. But we should be able to somehow make that safe, and if it's via `unsafe` code called from a macro that prevents users from doing silly things there (like passing a literal `42` as offset).

**Comment by ebassi** [comment]
> That sounds unexpected. @ebassi can you confirm that `get_template_child()` requires things to be bound first somehow?

It's entirely expected: how does `gtk_widget_get_template_child()` know what widget to instantiate and add to the hash table of children, otherwise? You need to call `gtk_widget_class_bind_template_child_full()` to tell which child widgets should be accessible from the template. [The documentation is pretty clear](https://developer.gnome.org/gtk3/stable/GtkWidget.html#gtk-widget-class-bind-template-child-full), but we can always improve it.

What is not needed is to set the offset inside the instance structure, or the private instance data structure; you can pass 0 for that, and GTK will not try to assign the child instance to a C field at the given offset from the beginning of the widget's instance data structure. This will make `gtk_widget_get_template_child()` work.

The only thing you need are:

 - a pointer to the class structure, so you can call `gtk_widget_class_bind_template_child_full()`
 - the name of the widget in the template
 - whether the child is an internal widget or not

**Comment by cyclopsian** [comment]
@ebassi You're right, thank you. I just tested it with `get_template_child` and the offset set to 0 and it works. That way `OnceCell` still can be used as intended. I updated my gist with the code I'm using now. Also there is a branch here that improves `gtk::Widget::get_template_child`: [get_template_child_0.8.1](https://github.com/gtk-rs/gtk/compare/0.8.1...cyclopsian:get_template_child_0.8.1)

@sdroege 
>Great, I'd suggest as a first step to add something along the lines of what I wrote above to the bindings. Something to hook into instance_init() so that this can be automatically called, and safe bindings for gtk_widget_class_set_template() (and the resource variant of it).

I am having trouble figuring out how to integrate this. Would it make sense to add some function to `ObjectSubclass` to have it store a `Fn` which is then called in `instance_init`? The idea would be that `WidgetSubclass` would then be able to call it.

>That's correct, and unfortunately it also does not work with a GClosure but directly with a function pointer so we can't do any runtime checks like with the GtkBuilder API either. In GTK4 this works completely different though, so hopefully less of a problem there.

Do you have any more details on this? It looks like language bindings are supposed to implement GtkBuilderScope, but that doesn't seem to have a way to get at the underlying object/signal from the builder to verify that it matches the closure?

>That actually looks quite nice and like something we could get in as a second step after the above. I'm not 100% sure about the safety of the bind_template_child_private() part, I'll review that in more detail once there's a MR. But we should be able to somehow make that safe, and if it's via unsafe code called from a macro that prevents users from doing silly things there (like passing a literal 42 as offset).

The `field_offset` crate makes it safe to get the offsets, it doesn't allow random values. The field has to exist and be of the correct type. But even with that it's still possible to bind a widget to the wrong type because the type is not known until the XML is read at runtime. It won't segfault, but will panic when calling `unwrap()` on the `TemplateWidget`.

I think I am going to scrap that idea though. At the moment I am preferring to call `get_template_child()` several times in `constructed()` and use that to populate a `OnceCell<WindowWidgets>` struct like in the subclassing example, because that way all the widgets get type-checked upon object construction. It may not make much of a difference though? Maybe it makes sense to support both methods.

**Comment by sdroege** [comment]
> Also there is a branch here that improves `gtk::Widget::get_template_child`: [get_template_child_0.8.1](https://github.com/gtk-rs/gtk/compare/0.8.1...cyclopsian:get_template_child_0.8.1)

Can you send a PR for that? You probably also want to make it a `Result` so that type mismatches can be distinguished from the non-existence of the widget.

> I am having trouble figuring out how to integrate this. Would it make sense to add some function to `ObjectSubclass` to have it store a `Fn` which is then called in `instance_init`? The idea would be that `WidgetSubclass` would then be able to call it.

Not really as `ObjectSubclass` is something that actual subclass implementors are using. Ideally we want to have something inside `gtk/src/subclass/widget.rs` that does this automatically, and the way to do that (for example) would be by having a new function as part of `IsSubclassable<T>`.

What problem did you run into there, what exactly did you try? If you submit a WIP PR for that I can try helping :)



> Do you have any more details on this? It looks like language bindings are supposed to implement GtkBuilderScope, but that doesn't seem to have a way to get at the underlying object/signal from the builder to verify that it matches the closure?

I don't know how exactly it's supposed to work. But the `GtkBuilderScope` interface has a vfunc that returns a `GClosure`, which would then allow doing type checks at runtime to prevent undefined behaviour and just abort the process.

The signal machinery already works based on `GClosure`s internally.

> The `field_offset` crate makes it safe to get the offsets, it doesn't allow random values. The field has to exist and be of the correct type. But even with that it's still possible to bind a widget to the wrong type because the type is not known until the XML is read at runtime. It won't segfault, but will panic when calling `unwrap()` on the `TemplateWidget`.

You could point it to any kind of field, even a `bool` for example, or not?

If we could enforce that it must be an offset to a `TemplateWidget<T>` then we can do all this safely and do the checks at runtime when trying to access the contained widget. I think that would be a nice API.

With `OnceCell` you always have to handle the possibility that it was not initialized yet whenever you try to use it, which makes it a bit annoying to use. I like the idea of the `TemplateWidget`.

In any case, I think we should move forward here step by step with smaller PRs :) a) `get_template_widget()`, b) `set_template***()` on the class, c) automatic initialization of the template in instance init, d) automatic widget binding like above.

I'd be happy to review and help with those once there are PRs.

**Comment by sdroege** [comment]
@cyclopsian Any news from your side, new questions, problems, ...? :)

**Comment by cyclopsian** [comment]
I think I need to upgrade my PC, both `rls` and `rustc` are really unbearably slow on an X200

**Comment by sdroege** [comment]
> I think I need to upgrade my PC, both `rls` and `rustc` are really unbearably slow on an X200

This might become a bit better with https://github.com/gtk-rs/gtk/pull/1007 (also rust-analyzer is faster and less resource hungry than rls).

**Comment by ebassi** [comment]
> Also there is a branch here that improves `gtk::Widget::get_template_child`: [get_template_child_0.8.1](https://github.com/gtk-rs/gtk/compare/0.8.1...cyclopsian:get_template_child_0.8.1)

I missed this at the time, but the change is wrong. The `get_template_child()` method of GtkWidget can be used to return *any object* created by a template, not just widgets. For instance, you could use `get_template_child()` to retrieve a `GtkAdjustment` set as the value of a property.
